File: android-qr-data-transfer/src/main/java/com/prgpascal/qrdatatransfer/BaseActivity.java
package com.prgpascal.qrdatatransfer;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.widget.Toast;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

public abstract class BaseActivity extends FragmentActivity {
    private static final int PERMISSIONS_REQUEST = 111;
    private static final String[] PERMISSIONS = new String[]{
            Manifest.permission.CAMERA
    };

    private boolean mIsResolvingPermissionRequests;

    protected void checkAppPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            ArrayList<String> requiredPermissions = new ArrayList<>();
            for (String permission : PERMISSIONS) {
                if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                    requiredPermissions.add(permission);
                }
            }

            if (requiredPermissions.size() > 0) {
                if (!mIsResolvingPermissionRequests) {
                    mIsResolvingPermissionRequests = true;
                    ActivityCompat.requestPermissions(this, requiredPermissions.toArray(new String[0]), PERMISSIONS_REQUEST);
                }
            } else {
                permissionsGranted();
            }
        } else {
            permissionsGranted();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSIONS_REQUEST) {
            mIsResolvingPermissionRequests = false;
            boolean allPermissionsGranted = false;

            if (grantResults.length > 0) {
                allPermissionsGranted = true;
                for (int permission : grantResults) {
                    if (permission != PackageManager.PERMISSION_GRANTED) {
                        allPermissionsGranted = false;
                        break;
                    }
                }
            }

            if (!allPermissionsGranted) {
                Toast.makeText(getApplicationContext(), getString(R.string.aqrdt_error_no_permissions), Toast.LENGTH_SHORT).show();
                finish();
            } else {
                permissionsGranted();
            }
        }
    }

    public abstract void permissionsGranted();

}

File: android-qr-data-transfer/src/main/java/com/prgpascal/qrdatatransfer/activities/BaseActivity.kt
package com.prgpascal.qrdatatransfer.activities

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.prgpascal.qrdatatransfer.R
import java.util.*

abstract class BaseActivity : FragmentActivity() {
    companion object {
        private const val PERMISSIONS_REQUEST = 111
        private val PERMISSIONS = arrayOf(
                Manifest.permission.CAMERA
        )
    }

    private var mIsResolvingPermissionRequests = false

    protected fun checkAppPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val requiredPermissions = ArrayList<String>()
            for (permission in PERMISSIONS) {
                if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                    requiredPermissions.add(permission)
                }
            }
            if (requiredPermissions.size > 0) {
                if (!mIsResolvingPermissionRequests) {
                    mIsResolvingPermissionRequests = true
                    ActivityCompat.requestPermissions(this, requiredPermissions.toTypedArray(), PERMISSIONS_REQUEST)
                }
            } else {
                permissionsGranted()
            }
        } else {
            permissionsGranted()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSIONS_REQUEST) {
            mIsResolvingPermissionRequests = false
            var allPermissionsGranted = false
            if (grantResults.isNotEmpty()) {
                allPermissionsGranted = true
                for (permission in grantResults) {
                    if (permission != PackageManager.PERMISSION_GRANTED) {
                        allPermissionsGranted = false
                        break
                    }
                }
            }
            if (!allPermissionsGranted) {
                Toast.makeText(applicationContext, getString(R.string.aqrdt_error_no_permissions), Toast.LENGTH_SHORT).show()
                finish()
            } else {
                permissionsGranted()
            }
        }
    }

    abstract fun permissionsGranted()

}